### ISSUE TITLE
SourceTextComparer's has odd IEqualityComparer semantics

### ISSUE THREAD

**Issue by ToddGrun** [dotnet/roslyn#71006]
The Equals method is using the content, whereas the GetHashCode method is using the contents and encoding. This mismatch seems like it could cause issues downstream.

**Comment by CyrusNajmabadi** [comment]
Yes.  THere are definite issues here.  This means you can generate two source-text instances that have the following behavior:

1. Each produces a different hash code for SourceTextComparer.GetHashCode
2. They return .Equals for SourceTextComparer.Equals

This violates the contract of IEqualityComparer.

The way to do this is to create the source-texts from byte[]s or streams, where one as a BOM and the other doesn't (but otherwise the same content).  For both, just pass the Encoding.Utf8 encoding.  This will produce source-texts with equal content, but different hashes (since the hashes use the 'checksum' of the ST, which uses the original bytes to compute itself).

There are two solutions here:

1. SourceTextComparer.GetHashCode should use SourceText.GetContentHash to produce the final hashcode (and that alone).  THis will being .GetHashCode in line with .Equals. This would make it so that SourceTextComparer has a *strictly content based* view of source-texts.  Two source texts with the same content would be considered the same, regardless of encoding, hash-algorithm, or originating bytes. 
2. SourceTextComparer.Equals should use the Encoding and Checksum to determine equality.  This will bring .Equals in line with .GetHashCode.  This will make it so that SourceTextComparer has a *strictly originating* based view of source-texts.  So for two source texts to be the same, they would have had to come from the same original bytes as well.

I don't acually know which is preferred for the use cases where SOurceTextComparer is used (looks like it's in incremental generators for example).

--

My preference is that we take teh former view of things (a content-based view).  It seems to me that if something is producing a source-text with the same content that that should be considered a no-op change for something like incremental-generators.

Anyways, this is onto compiler team to decide what to do.  I'm happy to share insights here though if there are any questions.

**Comment by CyrusNajmabadi** [comment]
Actually, just realized that for incremental-generators, we won't use GetHashCode at all.  It's just going to use .Equals.  So that shouldn't change.  

The only place i can see using the GetHashCode part of this is SourceTextValueProvider (which is used to cache data against a source-text within an analyzer).

But here's the crazy thing.  The STVP uses a CWT to cache data against the source-text (which is my expectation).  So this IEqualityComaprer is not used for that purpose *either*.

So i legit cannot tell what purpose this serves *at all*.  I think we likely can remove this type.

**Comment by CyrusNajmabadi** [comment]
@mavasani @chsienki to look at this.  I'm fairly certain this guy can be removed.  And in the case of SOurceTExts, we should be using *identity* semantics for analyzers, and for incremental-generators, we use content-equality and we *throw* on GetHashCode (since that should never be called).

**Comment by mavasani** [comment]
> The STVP uses a CWT to cache data against the source-text (which is my expectation). So this IEqualityComaprer is not used for that purpose either.

@CyrusNajmabadi The SourceTextValueProvider is not used for the CWT. It flows down to here: https://sourceroslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/AnalysisValueProvider.cs,e57a61365c6f659c,references, which is used as the comparer for the dictionary stored here: https://sourceroslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/CompilationAnalysisValueProvider.cs,25

I believe the core usage of a source/content based equality check over reference based equality check comes from this API: https://github.com/dotnet/roslyn/blob/2ddf1e71cb80c0db0809bc3c9b457ba368e98f94/src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs#L231-L244

This API allows analyzers to cache computed data from source texts, and the cache can live across compilations to avoid recomputations if the underlying content of the SourceText is identical between compilations even though the reference equality doesn't hold anymore. This only applies for IDE typing scenarios where analyzer instances (and hence the input SourceTextValueProvider instance passed down to TryGetValue API) can live across compilations.

> My preference is that we take teh former view of things (a content-based view)

This seems reasonable to me as well. Unless there is some analyzer out there that uses the above `TryGetValue` API and whose compute function also relies on the encoding of the source text for the eventual computed value. This seems unlikely to me though, as such an analyzer would already be not working as expected due to the conflicting `Equals` and `GetHashCode` implementations of the default `SourceTextComparer` that Todd has identified.

My take on this is that removing `SourceTextComparer` type and switching to the reference equality comparer for source texts might lead to analyzer performance regressions and might be a risky change. I'd rather we fix the semantics of `SourceTextComparer` with one of the two suggestions you mentioned (content-based view OR content + encoding based view)

**Comment by CyrusNajmabadi** [comment]
Works for me.  Going with content-based seems like the sensible approach.

**Comment by chsienki** [comment]
Content-based seems sensible. FWIW I don't think this is used at all for generators. A couple of the fuzz tests use it but that's it.
